File: 1to150/64_minimalPathsum.java
/*
[[1 3 1]
 [1 5 1]
 [4 2 1]
 [4 5 2]]
 
很明显，走到每一格的sum基于前面从哪里走过来的，所以对于第一row和第一列，走到每一格的大小是固定的，因为规定了只能从上往下走或者从左往右走

对于  [1 3]
      [1 5], right bottom 5 那里来说， 走到他的这个位置可以从 3 往下走到它，或者从1往右走到他。 很明显 min(3 + 5, 1 + 5)选择6
      
      
1. state function, dp[m][n], 每一个位置代表走到这位置的sum是多少
2. init，对于dp[0][0] 来说 就是自己，然后第一行和第一列的第二位数开始，都是dp[i -1 ] + 自己
3. transfer function: dp[i][j]= min(上+自己， 左+自己)
4. return dp[i-1][j-1]
*/
class Solution {
    public int minPathSum(int[][] grid) {
        if (grid == null){
            return 0;
        }
        //1 state function, each grid represents the sum needed to go up to this grid
        int[][] dp = new int[grid.length][grid[0].length];
        
        //2 init, the first grid is just itself, and the rest of the first column and first row is the itself + dplast
        dp[0][0] = grid[0][0];
        for (int i = 1; i < grid.length; i++){
            dp[i][0] = dp[i - 1][0] + grid[i][0];
        }
        
        for(int j = 1; j <grid[0].length; j++){
            dp[0][j] = dp[0][j - 1] + grid[0][j];
        }
        
        //transfer function
        for (int i = 1; i < grid.length; i++){
            for (int j = 1; j < grid[0].length; j++){
                int itself = grid[i][j];
                dp[i][j] = Math.min(dp[i - 1][j] + itself, dp[i][j - 1] + itself);
            }
        }
        
        return dp[grid.length - 1][grid[0].length - 1];
    }
}

File: 1to150/64_minimalPathsum.py
'''
1 state function: dp[m][n], 每一个位置代表了走到这个位置的时候，最小的path runningsum为多少
2 init function：第一row只能从左边走来，所以都是runningSum之前一位
                 第二咧同理
                 
3. transfer fuction
  1 -> 4 -> 5
  |
  2    ?            这个位置我们要选取最小的走法 所以 min(dp[i - 1][j】 + 自己， dp[i][j - 1] + 自己)
  |
  6
  
  但如果存在obstacle的情况，我们需要判断 dp[i - 1][j] 和 dp[i][j -1]是否为0， 如果为零说明有一条路不通， 那么就只加不为零的那部分/ 如果自己这个grid存在obtacle，那直接这里设为0 

'''


class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        
        if not grid:
            return 0
        
        mRows= len(grid)
        nCols = len(grid[0])
        
        dp = [[0 for j in range(nCols)] for i in range(mRows)]
        
        #init
        for i in range(mRows):
            if i == 0:
                dp[i][0] = grid[0][0]
                continue
            dp[i][0] += dp[i - 1][0] + grid[i][0]
            
        for j in range(nCols):
            if j == 0:
                dp[0][j] = grid[0][0]
                continue
            dp[0][j] += dp[0][j - 1] + grid[0][j]
            
        for i in range(1, mRows):
            for j in range(1, nCols):
                dp[i][j] = min(dp[i - 1][j] + grid[i][j], dp[i][j - 1] + grid[i][j])
                
        return dp[mRows - 1][nCols - 1]
            
        